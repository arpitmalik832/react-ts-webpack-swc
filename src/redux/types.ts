import {
  BaseQueryFn,
  TypedUseQueryStateResult,
} from '@reduxjs/toolkit/query/react';
import { AxiosError, AxiosInstance } from 'axios';

interface SampleApiResponse {
  data: string[];
}

export interface BaseQueryParams {
  axiosInstance: AxiosInstance;
  url: string;
}

export type SampleQueryResponse = TypedUseQueryStateResult<
  SampleApiResponse,
  AxiosError,
  BaseQueryFn<BaseQueryParams>
>;
