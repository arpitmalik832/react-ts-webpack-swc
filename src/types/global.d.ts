interface Window {
  backPress: () => void;
}

declare module '*.svg' {
  export const ReactComponent: React.FunctionComponent<
    React.SVGProps<SVGSVGElement>
  >;
  const src: string;
  export default src;
}

declare module '*.png' {
  const value: string;
  export default value;
}

declare module '*.jpg' {
  const value: string;
  export default value;
}

declare module '*.jpeg' {
  const value: string;
  export default value;
}

declare module '*.gif' {
  const value: string;
  export default value;
}

declare module '*.bmp' {
  const value: string;
  export default value;
}

declare module '*.tiff' {
  const value: string;
  export default value;
}

declare module '*.module.scss' {
  const content: { [className: string]: string };
  export default content;
}
