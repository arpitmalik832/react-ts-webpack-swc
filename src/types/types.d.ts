import type {
  InternalAxiosRequestConfig,
  AxiosResponse,
  AxiosInstance,
} from 'axios';
import type { configureStore } from '@reduxjs/toolkit';

interface VoidFunction extends VoidFunction {
  (...args: any[]): void;
}

interface ComponentWithSuspenseProps {
  component: React.ReactNode;
  fallback?: React.ReactNode;
}

interface HtmlContentProps {
  title: string;
  description: string;
}

interface ReduxProviderProps {
  children?: React.ReactNode;
  store: ReturnType<typeof configureStore>;
}

interface IconProps {
  name: string;
}

type AbortControllers = Record<string, AbortController>;

interface RequestMetadata {
  startTime: Date;
  endTime: Date;
  responseTime: number | bigint;
}

interface AppRedux {
  theme: string;
}

interface ApisRedux {
  api1Host: string;
  api1Headers: Record<string, string>;
  api1AxiosInstance?: AxiosInstance;
}

interface NavigationRedux {
  stack: Array<VoidFunction>;
}

interface ReduxState {
  app: AppRedux;
  apis: ApisRedux;
  navigation: NavigationRedux;
}

type ColorTheme = 'light' | 'dark';
type ColorType = 'background';
type ColorInnerType = 'primary';
type PrimitiveColor = 'black' | 'white';
type ColorSemanticLabel = '50' | '100' | '900';
type HexColor = `#${string & { length: 6 | 8 }}`;
type TypographyScale = 'h1';
type TypographyScaleValue = '*px';
type TypographyWeight = 'bold';
type TypographyWeightValue = number;

interface DesignTokens {
  'color-primitives': Record<
    PrimitiveColor,
    Record<
      ColorSemanticLabel,
      {
        type: 'value';
        color: HexColor;
      }
    >
  >;
  'color-semantics': Record<
    Theme,
    Record<
      Type,
      Record<
        InnerType,
        Record<
          ColorSemanticLabel,
          {
            type: 'value';
            color: HexColor;
          }
        >
      >
    >
  >;
  typography: {
    scale: Record<
      TypographyScale,
      Record<
        'font-size' | 'line-height' | 'letter-spacing',
        Record<'value', TypographyScaleValue>
      >
    >;
    weight: Record<TypographyWeight, Record<'value', TypographyWeightValue>>;
  };
}

type MQEventListener = (this: MediaQueryList, ev: MediaQueryListEvent) => void;

type EventListener = (this: Window, ev: Event) => void;

type BeforeUnloadEventListener = (this: Window, ev: BeforeUnloadEvent) => void;
