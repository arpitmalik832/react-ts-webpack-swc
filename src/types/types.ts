import type {
  AxiosInstance,
} from 'axios';

export type AllParams =
  | string
  | string[]
  | object
  | object[]
  | number
  | number[]
  | bigint
  | bigint[]
  | boolean
  | boolean[];

export interface VoidFunctionWithParams extends VoidFunction {
  (...args: AllParams[]): void;
}

export interface KeyValuePair<T> {
  key: string;
  value: T | Record<string, T>;
}

export interface AppRedux {
  theme: string;
}

export interface NavigationRedux {
  stack: VoidFunctionWithParams[];
}

export interface APIData {
  host: string;
  headers: Record<string, string>;
  axiosInstance: AxiosInstance;
}

export type ApisRedux = APIData[];

export interface ReduxState {
  app: AppRedux;
  apis: ApisRedux;
  navigation: NavigationRedux;
}