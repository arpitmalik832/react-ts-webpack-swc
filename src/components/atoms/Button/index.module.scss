.button {
  font-size: 12px;
  color: var(--background-primary-900);
  background-color: var(--background-primary-100);
  border: 1px solid var(--background-primary-900);
}
