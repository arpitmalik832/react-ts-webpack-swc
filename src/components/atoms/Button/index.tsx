import { slices } from '@arpitmalik832/react-js-rollup-monorepo-library';
import { useDispatch } from 'react-redux';

import s from './index.module.scss';

function Button() {
  const dispatch = useDispatch();

  function onButtonClick() {
    dispatch(slices.updateStore({ key: 'x', value: 'a' }));
  }

  return (
    <button
      type="button"
      data-testid="button"
      data-cy="button"
      className={s.button}
      onClick={onButtonClick}
    >
      Button
    </button>
  );
}

export default Button;
