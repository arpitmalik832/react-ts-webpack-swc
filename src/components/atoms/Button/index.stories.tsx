import { ReduxProvider } from '@arpitmalik832/react-js-rollup-monorepo-library';

import Button from './index';
import store from '../../../redux/store/store';

export default {
  title: 'Atoms/Button',
  component: () => (
    <ReduxProvider store={store}>
      <Button />
    </ReduxProvider>
  ),
  tags: ['autodocs'],
};

export const Primary = {
  args: {
    primary: true,
    label: 'Button',
  },
};
