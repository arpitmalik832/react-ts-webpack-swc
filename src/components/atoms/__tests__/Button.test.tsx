import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ReduxProvider, slices } from '@arpitmalik832/react-ts-rollup-library';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';

import Button from '../Button';
import { AppRedux } from '../../../types/types';

describe('Button unit tests', () => {
  let store: EnhancedStore<{
    app: AppRedux;
  }>;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        app: slices?.appSlice?.reducer,
      },
    });
    // Mock the dispatch function
    store.dispatch = jest.fn();
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  const renderButton = () =>
    render(
      <ReduxProvider store={store}>
        <Button />
      </ReduxProvider>,
    );

  test('Button snapshot test', () => {
    const { container } = renderButton();
    expect(container).toMatchSnapshot();
  });

  it('renders with correct text content', () => {
    const { getByTestId } = renderButton();
    expect(getByTestId('button')).toHaveTextContent('Button');
  });

  it('has correct button attributes', () => {
    const { getByTestId } = renderButton();
    const button = getByTestId('button');

    expect(button).toHaveAttribute('type', 'button');
    expect(button).toHaveAttribute('data-cy', 'button');
    expect(button).toHaveClass('button');
  });

  it('dispatches correct action on click', () => {
    const { getByTestId } = renderButton();
    const button = getByTestId('button');

    fireEvent.click(button);

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      slices.updateStore({ key: 'x', value: 'a' }),
    );
  });
});
