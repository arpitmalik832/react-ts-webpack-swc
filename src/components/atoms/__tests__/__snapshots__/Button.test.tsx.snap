// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button unit tests Button snapshot test 1`] = `
<div>
  <button
    class="button"
    data-cy="button"
    data-testid="button"
    type="button"
  >
    Button
  </button>
</div>
`;
